rust/lance/src/io/commit/conflict_resolver.rs

```diff
@@ -0,0 +1,722 @@
+                // 3. Write out new deletion files with existing deletes | affected rows.
+                // 4. Update the transaction with the new deletion files.
+
+                let fragments_ids_to_rewrite = self
+                    .initial_fragments
+                    .iter()
+                    .filter_map(|(_, (fragment, needs_rewrite))| {
+                        if *needs_rewrite {
+                            Some(fragment.id)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                // We are rewriting the deletion files on the *current* dataset.
+                let files_to_rewrite = dataset
+                    .fragments()
+                    .as_slice()
+                    .iter()
+                    .filter_map(|fragment| {
+                        if fragments_ids_to_rewrite.contains(&fragment.id) {
+                            Some((fragment.id, fragment.deletion_file.clone()))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                let existing_deletion_vecs = futures::stream::iter(files_to_rewrite)
+                    .map(|(fragment_id, deletion_file)| async move {
+                        read_deletion_file_cached(
+                            fragment_id,
+                            &deletion_file.expect("there should be a deletion file"),
+                            &dataset.base,
+                            dataset.object_store(),
+                            &dataset.session.file_metadata_cache,
+                        )
+                        .await
+                        .map(|dv| (fragment_id, dv))
+                    })
+                    .buffered(dataset.object_store().io_parallelism())
+                    .try_collect::<Vec<_>>()
+                    .await?;
+
+                // Check for row-level conflicts
+                let mut existing_deletions = RowIdTreeMap::new();
+                for (fragment_id, deletion_vec) in existing_deletion_vecs {
+                    existing_deletions
+                        .insert_bitmap(fragment_id as u32, deletion_vec.as_ref().into());
+                }
+                let conflicting_rows = existing_deletions.clone() & affected_rows.clone();
+                if conflicting_rows.len().map(|v| v > 0).unwrap_or(true) {
+                    let sample_addressed = conflicting_rows
+                        .row_ids()
+                        .unwrap()
+                        .take(5)
+                        .collect::<Vec<_>>();
+                    return Err(crate::Error::RetryableCommitConflict {
+                        version: dataset.manifest.version,
+                        source: format!(
+                            "Found conflicts for row addresses {:?}",
```

Comment:
I worry this error message could be confusing to users.  Should we use some variation of the other one:

```
This {} transaction was preempted by concurrent transaction {} at version {} (both modified rows at addresses {:?}). Please retry
```

Reply:
updated